### PR TITLE
Introducing ur3

### PR DESCRIPTION
**Pull request by fmessmer**
includes #209, #210 and #211 

this PR is introducing urdf, launch and config files for the UR3 robot

It is not yet finalized:
- meshes could use some color
- Parameters: DH, Center-of-Gravity, Masses should be retrieved from a ur.conf file from a UR3 controller

It would be appreciated if someone could assist in finalizing the URDF and the meshes...
...as well as test it with the hardware!
I'm all happy to see this draft PR superseded with another PR finalizing the UR3!

I'll create a `ur3_moveit_config` and adjust `ur_kinematics` to support the UR3 as soon as the `ur_description` has been finalized!

### CODE REVIEW

File: ur_description/urdf/ur3.urdf.xacro [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+  Author: Felix Messmer
+-->
+
+  <xacro:include filename="$(find ur_description)/urdf/ur.transmission.xacro" />
+  <xacro:include filename="$(find ur_description)/urdf/ur.gazebo.xacro" />
+
+  <property name="pi" value="3.14159265" />
+
+  <!-- Inertia parameters -->
+  <!-- ToDo: get values for UR3 -->
```
Comment (ipa-fxm) [inline ur_description/urdf/ur3.urdf.xacro]: get this from ur.conf

File: ur_description/urdf/ur3.urdf.xacro [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+  <property name="base_mass" value="4.0" />  <!-- This mass might be incorrect -->
+  <property name="shoulder_mass" value="2.0" />
+  <property name="upper_arm_mass" value="5.0" />
+  <property name="forearm_mass" value="1.5" />
+  <property name="wrist_1_mass" value="1.0" />
+  <property name="wrist_2_mass" value="1.0" />
+  <property name="wrist_3_mass" value="0.1" />
```
Comment (fmessmer) [inline ur_description/urdf/ur3.urdf.xacro]: get these values from a ur.conf

File: ur_description/urdf/ur3.urdf.xacro [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+       but are not verified for the correct permutation.
+       The permutation was guessed by looking at the UR5 parameters.
+       Serious use of these parameters needs further inspection. -->
+  <!-- ToDo: get values for UR3 -->
+  <property name="shoulder_cog" value="0.00008 0.00244 -0.037" />
+  <property name="upper_arm_cog" value="0.00001 0.15061 0.38757" />
+  <property name="forearm_cog" value="-0.00012 0.06112 0.1984" />
+  <property name="wrist_1_cog" value="-0.00021 -0.00112 0.02269" />
+  <property name="wrist_2_cog" value="-0.00021 0.00112 0.002269" />
+  <property name="wrist_3_cog" value="0 -0.001156 -0.00149" />
```
Comment (fmessmer) [inline ur_description/urdf/ur3.urdf.xacro]: get these values from a ur.conf

File: ur_description/urdf/ur3.urdf.xacro [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+
+  <!-- Kinematic model -->
+  <!-- Properties from urcontrol.conf -->
+  <!-- ToDo: get values for UR3 -->
+  <property name="ur3_d1" value="0.15" />
+  <property name="ur3_a2" value="-0.244" />
+  <property name="ur3_a3" value="-0.2124" />
+  <property name="ur3_d4" value="0.1146" />
+  <property name="ur3_d5" value="0.0878" />
+  <property name="ur3_d6" value="0.0835" />
```
Comment (fmessmer) [inline ur_description/urdf/ur3.urdf.xacro]: get these values from a ur.conf

File: ur_description/urdf/ur3.urdf.xacro [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+
+  <!-- Arbitrary offsets for shoulder/elbow joints -->
+  <!-- ToDo: get values for UR3 -->
+  <property name="shoulder_offset" value="0.121" />  <!-- measured from model -->
+  <property name="elbow_offset" value="-0.0937" /> <!-- measured from model -->
```
Comment (fmessmer) [inline ur_description/urdf/ur3.urdf.xacro]: get these values from a ur.conf

**Comment by gavanderhoorn** [comment]
@ipa-fxm: from your own comments it would seem we'd have to get those values from somewhere before merging this, right?

Recent `ursim` tarballs contain a `urcontrol.conf.UR3` which we could probably use.

**Comment by gavanderhoorn** [comment]
> It would be appreciated if someone could assist in finalizing the URDF and the meshes [..]

What kind of assistance are you looking for, exactly?

**Comment by fmessmer** [comment]
@gavanderhoorn 
Yes, I received a urcontrol.conf for the UR3 already, but I haven't had the time to apply the correct values to the PR...

In particular, I'd be happy if someone would find the time to prepare the *.stl and *.dae mesh files...:wink:...as most likely I won't find the time until end of July....

**Comment by jeppewalther** [comment]
Hi i am starting a project at university next week. Where we will use a UR3. I will be happy to do some work to get this running with ROS If you guys can give me some pointers on how to get started. I am a bit of a novice but did some previous work to support older version of CB5 https://github.com/ros-industrial/universal_robot/pull/153

**Comment by fmessmer** [comment]
Finally, I found the time to fix the remaining issues with the DH parameters and the meshes...
The URDF now should reflect the proper UR3 kinematics.

@jeppewalther 
I'd appreciate if you could review and test the related PR's!

Use #214 as it includes all PRs #209, #210, #211, #212, #213 and #214

**Comment by jeppewalther** [comment]
I will be happy to test it as soon as i get my hands on the hardware.

**Comment by jeppewalther** [comment]
What version have you implemented with. I am currently running with the newest one. Simulation only the hardware is delayed. 

Polyscope: 3.1.18024
Robot Controller 3.1.18024

**Comment by fmessmer** [comment]
In fact, I only implemented the URDF, launch/config files and moveit_config without testing it with real hardware or URSim....I just created the model and prepared everything to be used...

According to [this comment](https://github.com/ros-industrial/universal_robot/issues/199#issuecomment-127335599), either the problem with TCP connection or still a UR protocol incompatibility with ur_driver is the problem...

Unfortunately, I won't be able to look into this before September....

**Comment by jeppewalther** [comment]
I got URSim running with Moveit. I have not experienced the problems with TCP connection. We should get the UR3 next week.

**Comment by fmessmer** [comment]
Oh, cool!
And you used #214, i.e. [my branch](https://github.com/ipa-fxm/universal_robot/tree/ur3_moveit_config)?

Were any changes in  ur_driver required? E.g. from [this branch](https://github.com/ipa-fxm/universal_robot/tree/indigo_fix_deserialize_v3)?
Just wondering how this could work as I thought they changed the protocol....

@jeppewalther  Please report in case additional changes had to be applied to get it working...

**Comment by jeppewalther** [comment]
Yes i am using #214 . I have not changed any code. I am amble to controle the robot in URSim with the Moveit GUI in RVIZ. 
shot

I might have been too quick in declaring it working. I just assumed that everything was working when I could move the robot. 

I looked around and there is an error that I did not see at first. (below) 

```
roslaunch ur_bringup ur3_bringup.launch limited:=true robot_ip:=localhost
... logging to /home/juw/.ros/log/4902298e-4378-11e5-abf6-ac9e17b769e1/roslaunch-d-main-8399.log
Checking log directory for disk usage. This may take awhile.
Press Ctrl-C to interrupt
Done checking log file disk usage. Usage is <1GB.

started roslaunch server http://d-main:51209/

SUMMARY
========

PARAMETERS
 * /robot_description: <?xml version="1....
 * /robot_ip_address: localhost
 * /robot_reverse_port: 50001
 * /rosdistro: indigo
 * /rosversion: 1.11.10
 * /tf2_buffer_server/buffer_size: 120.0
 * /ur_driver/max_payload: 3.0
 * /ur_driver/max_velocity: 10.0
 * /ur_driver/min_payload: 0.0

NODES
  /
    robot_state_publisher (robot_state_publisher/state_publisher)
    tf2_buffer_server (tf2_ros/buffer_server)
    ur_driver (ur_driver/driver.py)

auto-starting new master
process[master]: started with pid [8414]
ROS_MASTER_URI=http://localhost:11311

setting /run_id to 4902298e-4378-11e5-abf6-ac9e17b769e1
process[rosout-1]: started with pid [8427]
started core service [/rosout]
process[robot_state_publisher-2]: started with pid [8444]
process[ur_driver-3]: started with pid [8467]
process[tf2_buffer_server-4]: started with pid [8468]
Setting prefix to 
[WARN] [WallTime: 1439661961.482394] No calibration offset for joint "shoulder_pan_joint"
[WARN] [WallTime: 1439661961.482587] No calibration offset for joint "shoulder_lift_joint"
[WARN] [WallTime: 1439661961.482758] No calibration offset for joint "elbow_joint"
[WARN] [WallTime: 1439661961.482911] No calibration offset for joint "wrist_1_joint"
[WARN] [WallTime: 1439661961.483070] No calibration offset for joint "wrist_2_joint"
[WARN] [WallTime: 1439661961.483210] No calibration offset for joint "wrist_3_joint"
[INFO] [WallTime: 1439661961.483345] No calibration offsets loaded from urdf
[INFO] [WallTime: 1439661961.484048] Max velocity accepted by ur_driver: 10.0 [rad/s]
[INFO] [WallTime: 1439661961.485242] Bounds for Payload: [0.0, 3.0]
Disconnected.  Reconnecting
[INFO] [WallTime: 1439661961.510160] Programming the robot
[INFO] [WallTime: 1439661961.511805] Programming the robot at localhost
Likely a syntax error:
4���������  URControlFhJul 07 2015, 15:22:03
Exception in thread URConnection:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/home/juw/checkout/robot_printer/workspace/src/universal_robot/ur_driver/src/ur_driver/driver.py", line 279, in __run
    self.__on_packet(packet)
  File "/home/juw/checkout/robot_printer/workspace/src/universal_robot/ur_driver/src/ur_driver/driver.py", line 191, in __on_packet
    state = RobotState.unpack(buf)
  File "/home/juw/checkout/robot_printer/workspace/src/universal_robot/ur_driver/src/ur_driver/deserialize.py", line 324, in unpack
    raise Exception("Fatal error when unpacking RobotState packet")
Exception: Fatal error when unpacking RobotState packet

Handling a request
[2015-08-15 20:06:01.688310] Out: hello
[INFO] [WallTime: 1439661961.712105] Robot connected
The action server for this driver has been started
```

**Comment by jeppewalther** [comment]
I see now that is the exact same problem as in https://github.com/ros-industrial/universal_robot/issues/183 sorry for the confusion

**Comment by ashfaqfarooqui** [comment]
@jeppewalther That error is related to the new packet state introduced by UR, Use [this branch](https://github.com/ipa-fxm/universal_robot/tree/indigo_fix_deserialize_v3) and try connecting.

**Comment by jeppewalther** [comment]
Ok that seems to work better. I simply merged https://github.com/ipa-fxm/universal_robot/tree/indigo_fix_deserialize_v3 and #214 to get support for the UR3. 

I can move the robot several times with Moveit and get state updates from the URSim without crashing. However I did need to restart all the modules a few times.

**Pull request merged**